### PR TITLE
Feat: Rich vertical list pattern

### PR DESCRIPTION
**Pull request by jmuzina**
## Done

Builds macro, examples, and documentation for the rich vertical list pattern per [design](https://www.figma.com/design/Hll6ZRvHyTTGBZZSoRd8ei/Higher-order-components-(HOCs)?node-id=645-17687&m=dev) and [spec](https://docs.google.com/document/d/1axvJzepn3xgSKVrQJG6VhN6VGvoBpC5RT1r5-Py0dW8/edit?tab=t.0)

Fixes [WD-14038](https://warthogs.atlassian.net/browse/WD-14038)

## QA

- Review all [vertical rich list examples](https://vanilla-framework-5306.demos.haus/docs/examples/patterns/rich-list-vertical/combined) and compare them against [design](https://www.figma.com/design/Hll6ZRvHyTTGBZZSoRd8ei/Higher-order-components-(HOCs)?node-id=645-17687&m=dev)
- Review [rich list docs](https://vanilla-framework-5306.demos.haus/docs/patterns/rich-list-vertical)
- Review [4.18.0 release notes](https://vanilla-framework-5306.demos.haus/docs/whats-new)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [ ] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix release (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [x] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).

## Screenshots
Screenshot from 2024-08-16 14-57-46
Screenshot from 2024-08-16 14-57-53
Screenshot from 2024-08-16 14-58-03

[WD-14038]: https://warthogs.atlassian.net/browse/WD-14038?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by jmuzina** [comment]
> Looks like the combined demo link doesn't work

Fixed, it was just pointing to the wrong url :)

File: templates/_macros/rich-vertical-list.jinja
Comment (jmuzina) [reply templates/_macros/rich-vertical-list.jinja:33]: This is needed due to a quirk with how Jinja works - `List::append()` returns `None`, so when it's called in Jinja, it will render `None` unless you use an `or ""` or use `{% set _ = list.append(...) %}`.

Screenshot from 2024-08-16 17-09-50

This is also in [hero](https://github.com/canonical/vanilla-framework/pull/5189#discussion_r1684643581) ([src](https://github.com/canonical/vanilla-framework/blob/154e8584179e4d6fca56745e3c7d03fbb9526db1/templates/_macros/hero.jinja#L68))

File: templates/_macros/rich-vertical-list.jinja
Comment (jmuzina) [reply templates/_macros/rich-vertical-list.jinja]: Fixed

File: templates/docs/examples/patterns/rich-list/vertical/default.html
Comment (jmuzina) [reply templates/docs/examples/patterns/rich-list/vertical/default.html]: Fixed

**Comment by lyubomir-popov** [comment]
looks great overall! 1 small glitch I noticed: some blurry logos - can you please use the official ones uploaded here https://assets.ubuntu.com/manager?tag=variable-width-logos-july-2024

docs suggestions:

 - flipped: mention that this is useful when having multiple vertical rich lists in a sequence, alternating between flipped and default#

 - this feels too empty. Can we say we should have at least one of the optional elements, e.g. a paragraph or list?
 
image

**Comment by jmuzina** [comment]
> 1 small glitch I noticed: some blurry logos - can you please use the official ones uploaded here https://assets.ubuntu.com/manager?tag=variable-width-logos-july-2024

The only Intel logo I found here doesn't look quite right - it has a dotted border and is too wide. Do you have a better one?
Screenshot 2024-08-20 at 4 03 50 PM

Also, the mockup uses Intel logo twice in the same logo section - it doesn't really matter for the example's sake, but I figured I'd mention it in case there is another logo we could use

File: templates/_macros/rich-vertical-list.jinja
Comment (jmuzina) [reply templates/_macros/rich-vertical-list.jinja:3]: Just for clarification, are you looking for the integer variant approach as explored in #5304 ? If so, I think we should have a wider team discussion on this approach before committing to it

**Comment by lyubomir-popov** [comment]
demo not working

**Comment by jmuzina** [comment]
> demo not working

@lyubomir-popov I've requested the demo to restart in Jenkins, it should be up in a few minutes if things are working properly

**Comment by jmuzina** [comment]
Tomorrow I'll update the images on this to use the new responsive image container. They get quite tall on smaller screens, so we can take advantage of a wider aspect ratio in those cases to make things look better.

**Comment by jmuzina** [comment]
> Tomorrow I'll update the images on this to use the new responsive image container. They get quite tall on smaller screens, so we can take advantage of a wider aspect ratio in those cases to make things look better.

I don't think this is doable right now, as the example images get super cropped (because we are using `is-cover`) on smaller screens, and the images we are using don't look good when cropped.

Maybe if `is-cover` had `-on-medium`, `-on-large`, etc, this would work, but for now let's leave this PR as-is

File: templates/_macros/rich-vertical-list.jinja
Comment (jmuzina) [reply templates/_macros/rich-vertical-list.jinja]: Changed to `title_text`

File: templates/_macros/rich-vertical-list.jinja
Comment (bartaz) [reply templates/_macros/rich-vertical-list.jinja:3]: I don't want to be blocking this on the decision. We already have couple patterns merged that use flags for variants, so this follows existing convention.

Obviously, it would be good to have a conclusion before we officially publish macros (not to change API later), but I think we are fine doing any changes if needed before that.

File: templates/_macros/rich-vertical-list.jinja
Comment (bartaz) [inline templates/_macros/rich-vertical-list.jinja:6]: I'm a bit hesitant about implementing it like this.

Having multiple separate slots seems a bit mundane, especially that it seems you are required to put a specific element with a specific class name it it. Similar with list item below.

We are making strong assumptions there. Pattern provides the wrapping element of a component (`p-logo-section` or `p-list-divided`) and we are expecting content with class names of child element of these specific components. So we assume that developer understands that a particular component (that they are not using directly) is there. 

I don't think we should be doing that. 

We probably need to either have one slot per component and ask for a full specific component:

```
   {% if slot=="logo-section" %}
    <div class="p-logo-section">… </div>
   {% endif %}
```

or not require HTML at all, and only ask for contents of the list items, without wrapping element or class name.

```
   {% if slot=="list_item_1" %}
    Contents of list item without the element itself
   {% endif %}
```

But this wont work in case of logo images, and would be limited to a specific type of list (it will not be possible to change type of list or bullets).

Comment (jmuzina) [reply templates/_macros/rich-vertical-list.jinja]: Switched to using a single logo section slot.

Investigating best way to adjust the list items.

Comment (jmuzina) [reply templates/_macros/rich-vertical-list.jinja]: @bartaz 

I'm hesitant to make users pass in the list as its own element (it seems rather unwieldy, since the point of this pattern *is* the list, I'd like to try to abstract its markup away slightly if possible). This would also prevent us from placing limits on the  total number of list items.

I also don't think we should restrict people to a specific type of list. If I had to implement it in one of these two ways, I'd probably choose the first way (passing the entire list in).

We could also pass in the list as an array of JSON objects, although this is a bit of a departure from our previous param/slot conventions:

```html
{% call(slot) rich_vertical_list(
  title_text='H2 - this can take up to two lines, but best to stick to one line.',
  list_items=[
    {
      'classes': ['is-ticked'],
      'html': 'No mandatory subscriptions for Ubuntu'
    },
    {
      'classes': ['is-ticked'],
      'html': (
        '<a href="https://ubuntu.com/security">Security</a>, '
        '<a href="https://ubuntu.com/support">Support</a>, '
        'and fully managed '
        '<a href="https://ubuntu.com/openstack/managed">Infra</a> '
        'and <a href="https://ubuntu.com/managed/apps">Apps</a>'
      ),
    },
    {
      'classes': ['is-ticked'],
      'html': 'Public cloud, Data centre, Edge cluster and Appliances',
    }
  ]
) -%}
  {% set list_items = list_items[:15] %}
  {%- if list_items|length > 0 %}
  <!--Optional list-->
  <ul class="p-list--divided">
    {% for list_item in list_items %}
      <div class="p-list__item {{ (list_item.classes or []) | join(' ') }}">
        {{ list_item.html | safe }}
      </div>
   {% endfor %}
 </ul>
{%- endif -%}
 
{% endcall %}
```

@pastelcyborg - any opinion here?

Comment (jmuzina) [reply templates/_macros/rich-vertical-list.jinja]: Updated to passing in list item contents via slots, and setting list tick type using parameter.

**Comment by jmuzina** [comment]
The team has decided in SU today that we will be starting our HOC speccing efforts by retroactively speccing our to-be-delivered HOCs before they are merged. So, this PR will be blocked until review and approval of a spec by design, UX, and code.

**Comment by jmuzina** [comment]
@lyubomir-popov Tall images that are stacked vertically on small and medium breakpoints can disrupt the page flow a bit by making the user scroll past a very tall image. This is especially noticeable in the flipped variant.

To account for this, the examples now have two images - one with a wide aspect ratio for smaller screens, and one with a taller aspect ratio for larger screens.

Ex: before:
Screenshot 2024-10-23 at 11 37 37 AM

After:
Screenshot 2024-10-23 at 11 38 16 AM

Let me know if you think this is unnecessary or if there's a better way around it - just wanted to make sure these look as presentable as possible for release :)

**Review by bartaz — APPROVED** [review]
Looks good, thanks!

**Comment by mattea-turic** [comment]
> Quote reply

This looks good to me :)

**Pull request merged**
